shared/trust/hmac.go

```diff
@@ -0,0 +1,205 @@
+func (h *HMAC) Version() HMACVersion {
+	return h.conf.Version
+}
+
+// ParseHTTPHeader extracts the actual version and HMAC from the Authorization header.
+func (h *HMAC) ParseHTTPHeader(header string) (HMACVersion, []byte, error) {
+	version, hmacStr, err := h.splitVersionFromHMAC(header)
+	if err != nil {
+		return "", nil, err
+	}
+
+	hmac, err := hex.DecodeString(hmacStr)
+	if err != nil {
+		return "", nil, fmt.Errorf("Failed to decode the HMAC: %w", err)
+	}
+
+	return version, hmac, nil
+}
+
+// WriteBytes creates a new HMAC hash using the given bytes.
+func (h *HMAC) WriteBytes(b []byte) ([]byte, error) {
+	mac := hmac.New(h.conf.HashFunc, h.key)
+	_, err := mac.Write(b)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to create HMAC: %w", err)
+	}
+
+	return mac.Sum(nil), nil
+}
+
+// WriteJSON creates a new HMAC hash using the given struct.
+func (h *HMAC) WriteJSON(v any) ([]byte, error) {
+	payload, err := json.Marshal(v)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to marshal payload: %w", err)
+	}
+
+	return h.WriteBytes(payload)
+}
+
+// WriteRequest creates a new HMAC hash using the given request.
+// It will extract the requests body.
+func (h *HMAC) WriteRequest(r *http.Request) ([]byte, error) {
+	body, err := io.ReadAll(r.Body)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to read request body: %w", err)
+	}
+
+	defer func() {
+		// Reset the request body for the actual handler.
+		r.Body = io.NopCloser(bytes.NewBuffer(body))
+	}()
+
+	err = r.Body.Close()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to close the request body: %w", err)
+	}
+
+	return h.WriteBytes(body)
+}
+
+// Equal returns true in case hmac1 is identical to hmac2.
+func (h *HMAC) Equal(hmac1 []byte, hmac2 []byte) bool {
+	return hmac.Equal(hmac1, hmac2)
+}
+
+// HMACEqual extracts the HMAC from the Authorization header and
+// validates if it is equal to the HMAC created from the request's body
+// using the given formatter.
+func HMACEqual(hmac HMACFormatter, r *http.Request) error {
+	authHeader := r.Header.Get("Authorization")
+	if authHeader == "" {
+		return api.StatusErrorf(http.StatusBadRequest, "Authorization header is missing")
+	}
+
+	version, hmacFromHeader, err := hmac.ParseHTTPHeader(authHeader)
+	if err != nil {
+		return api.StatusErrorf(http.StatusBadRequest, "Failed to parse Authorization header: %w", err)
+	}
+
+	hmacVersion := hmac.Version()
+	if version != hmacVersion {
+		return api.StatusErrorf(http.StatusBadRequest, "Authorization header uses version %q but expected %q", version, hmacVersion)
+	}
+
+	hmacFromBody, err := hmac.WriteRequest(r)
+	if err != nil {
+		return api.StatusErrorf(http.StatusInternalServerError, "Failed to write request body: %w", err)
+	}
+
+	if !hmac.Equal(hmacFromHeader, hmacFromBody) {
+		return api.StatusErrorf(http.StatusForbidden, "Invalid HMAC")
+	}
+
+	return nil
+}
+
+// HMACAuthorizationHeader returns the HMAC on v using the given formatter.
```

Comment:
Returns the HMAC as an Authorization header?

Reply:
Fixed.